File: packages/interface/src/apdu/command-apdu.ts
/**
 * The CommandApdu class is for constructing and parsing APDU commands (standard and extended).
 * This class works in both Node.js and browser environments.
 */
export class CommandApdu {
  public readonly cla: number;
  public readonly ins: number;
  public readonly p1: number;
  public readonly p2: number;
  public readonly data: Uint8Array<ArrayBuffer> | null;
  public readonly le: number | null;

  /**
   * Constructor
   *
   * @param cla - CLA byte (0x00 to 0xFF)
   * @param ins - INS byte (0x00 to 0xFF)
   * @param p1 - P1 parameter (0x00 to 0xFF)
   * @param p2 - P2 parameter (0x00 to 0xFF)
   * @param data - Optional data payload (Uint8Array or null)
   * @param le - Optional expected response data length (number or null)
   */
  constructor(
    cla: number,
    ins: number,
    p1: number,
    p2: number,
    data: Uint8Array<ArrayBuffer> | null = null,
    le: number | null = null,
  ) {
    // Validate input bytes
    [cla, ins, p1, p2].forEach((byte, index) => {
      if (!Number.isInteger(byte) || byte < 0x00 || byte > 0xff) {
        throw new RangeError(
          `Byte ${index} (value: ${byte}) must be an integer between 0x00 and 0xFF.`,
        );
      }
    });

    this.cla = cla;
    this.ins = ins;
    this.p1 = p1;
    this.p2 = p2;
    this.data = data;
    this.le = le;
  }

  /**
   * Method to serialize CommandApdu to Uint8Array
   *
   * @returns Uint8Array representing the APDU command
   */
  public toUint8Array(): Uint8Array<ArrayBuffer> {
    const header = new Uint8Array([this.cla, this.ins, this.p1, this.p2]);
    const hasData = this.data !== null && this.data.length > 0;
    const dataLen = hasData ? this.data.length : 0;
    const isExtended =
      (hasData && dataLen > 255) || (this.le !== null && this.le > 256);

    let bodyLen = 0;
    if (hasData && this.le !== null) {
      bodyLen = isExtended ? 1 + 2 + dataLen + 2 : 1 + dataLen + 1;
    } else if (hasData) {
      bodyLen = isExtended ? 1 + 2 + dataLen : 1 + dataLen;
    } else if (this.le !== null) {
      bodyLen = isExtended ? 1 + 2 : 1;
    }

    const bodyBuffer = new ArrayBuffer(bodyLen);
    const view = new DataView(bodyBuffer);
    let offset = 0;

    if (hasData && this.le !== null) {
      if (isExtended) {
        view.setUint8(offset++, 0x00);
        view.setUint16(offset, dataLen);
        offset += 2;
        new Uint8Array(bodyBuffer).set(this.data, offset);
        offset += dataLen;
        view.setUint16(offset, this.le);
      } else {
        view.setUint8(offset++, dataLen);
        new Uint8Array(bodyBuffer).set(this.data, offset);
        offset += dataLen;
        view.setUint8(offset, this.le);
      }
    } else if (hasData) {
      if (isExtended) {
        view.setUint8(offset++, 0x00);
        view.setUint16(offset, dataLen);
        offset += 2;
        new Uint8Array(bodyBuffer).set(this.data, offset);
      } else {
        view.setUint8(offset++, dataLen);
        new Uint8Array(bodyBuffer).set(this.data, offset);
      }
    } else if (this.le !== null) {
      if (isExtended) {
        view.setUint8(offset++, 0x00);
        view.setUint16(offset, this.le);
      } else {
        view.setUint8(offset, this.le);
      }
    }

    const apdu = new Uint8Array(header.length + bodyLen);
    apdu.set(header, 0);
    apdu.set(new Uint8Array(bodyBuffer), header.length);
    return apdu;
  }

  /**
   * Method to serialize CommandApdu to a hexadecimal string
   *
   * @returns Hexadecimal string of the APDU command
   */
  public toHexString(): string {
    const byteArray = this.toUint8Array();
    return Array.from(byteArray)
      .map((byte) => byte.toString(16).padStart(2, "0"))
      .join("")
      .toUpperCase();
  }

  /**
   * Static method to parse a Uint8Array into a CommandApdu
   *
   * @param byteArray - Uint8Array to parse
   * @returns Parsed CommandApdu instance
   */
  public static fromUint8Array(
    byteArray: Uint8Array<ArrayBuffer>,
  ): CommandApdu {
    if (!(byteArray instanceof Uint8Array)) {
      throw new TypeError("Input must be a Uint8Array.");
    }

    if (byteArray.length < 4) {
      throw new RangeError("Input is too short to be a valid APDU command.");
    }

    const cla = byteArray[0];
    const ins = byteArray[1];
    const p1 = byteArray[2];
    const p2 = byteArray[3];
    let data: Uint8Array<ArrayBuffer> | null = null;
    let le: number | null = null;
    let index = 4;

    if (byteArray.length === 4) {
      // Case 1: Only CLA, INS, P1, P2
      return new CommandApdu(cla, ins, p1, p2);
    }

    if (byteArray.length === 5) {
      le = byteArray[4] === 0 ? 256 : byteArray[4];
      // Case 2: CLA, INS, P1, P2, Le
      return new CommandApdu(cla, ins, p1, p2, null, le);
    }

    /* If Command APDU is not case 2 and firstIndicator is 0x00, it is extended APDU */
    const firstIndicator = byteArray[index];
    if (firstIndicator !== 0x00) {
      // Case 3 or 4: CLA, INS, P1, P2, Lc, Data, [Le]
      const lc = byteArray[index];
      index += 1;

      if (byteArray.length === index + lc) {
        // Case 3: CLA, INS, P1, P2, Lc, Data
        data = byteArray.slice(index, index + lc);
        return new CommandApdu(cla, ins, p1, p2, data, null);
      } else if (byteArray.length === index + lc + 1) {
        // Case 4: CLA, INS, P1, P2, Lc, Data, Le
        data = byteArray.slice(index, index + lc);
        le = byteArray[index + lc];
        return new CommandApdu(cla, ins, p1, p2, data, le);
      } else {
        throw new RangeError("APDU structure is invalid.");
      }
    } else {
      // Extended APDU
      index += 1;
      if (byteArray.length < index + 2) {
        throw new RangeError(
          "For extended APDU, length of Lc or Le is insufficient.",
        );
      }

      // Read the next two bytes
      const lcOrLe = (byteArray[index] << 8) | byteArray[index + 1];
      index += 2;

      if (byteArray.length === index) {
        // Case 2 Extended: CLA, INS, P1, P2, 00, Le1, Le2
        le = lcOrLe === 0 ? 65536 : lcOrLe;
        return new CommandApdu(cla, ins, p1, p2, null, le);
      }

      const lc = lcOrLe;

      if (byteArray.length < index + lc) {
        throw new RangeError("Data is insufficient for the specified Lc.");
      }

      data = byteArray.slice(index, index + lc);
      index += lc;

      if (byteArray.length === index) {
        // Case 4 Extended: CLA, INS, P1, P2, 00, Data
        return new CommandApdu(cla, ins, p1, p2, data, null);
      }

      if (byteArray.length === index + 2) {
        // Case 4 Extended: CLA, INS, P1, P2, 00, Data, Le1, Le2
        le = (byteArray[index] << 8) | byteArray[index + 1];
        if (le === 0) {
          le = 65536;
        }
        return new CommandApdu(cla, ins, p1, p2, data, le);
      }
      throw new RangeError("Extended APDU structure is invalid.");
    }
  }

  /**
   * Method to return the APDU command as a human-readable hexadecimal string
   *
   * @returns Hexadecimal string of the APDU command
   */
  public toString(): string {
    return this.toHexString();
  }
}

File: packages/rn/android/src/main/java/expo/modules/jsapdu/apdu/CommandApdu.kt
package expo.modules.jsapdu.apdu

import expo.modules.jsapdu.ValidationError

/**
 * Command APDU (Application Protocol Data Unit)
 * Represents a command sent to a smart card
 */
class CommandApdu {
  val cla: Byte  // Class byte
  val ins: Byte  // Instruction byte
  val p1: Byte   // Parameter 1
  val p2: Byte   // Parameter 2
  val data: ByteArray?  // Command data
  val le: Int?   // Expected response length
  
  /**
   * Create a command APDU from individual components
   * @param cla Class byte
   * @param ins Instruction byte
   * @param p1 Parameter 1
   * @param p2 Parameter 2
   * @param data Command data (optional)
   * @param le Expected response length (optional)
   */
  constructor(
    cla: Int,
    ins: Int,
    p1: Int,
    p2: Int,
    data: ByteArray? = null,
    le: Int? = null
  ) {
    this.cla = cla.toByte()
    this.ins = ins.toByte()
    this.p1 = p1.toByte()
    this.p2 = p2.toByte()
    this.data = data
    this.le = le
    
    validate()
  }
  
  /**
   * Create a command APDU from a byte array
   * @param bytes APDU as byte array
   */
  constructor(bytes: ByteArray) {
    if (bytes.size < 4) {
      throw ValidationError(
        "APDU must be at least 4 bytes long",
        "bytes",
        bytes
      )
    }
    
    this.cla = bytes[0]
    this.ins = bytes[1]
    this.p1 = bytes[2]
    this.p2 = bytes[3]
    
    // Parse data and Le
    if (bytes.size == 4) {
      // Case 1: No data, no Le
      this.data = null
      this.le = null
    } else if (bytes.size == 5) {
      // Case 2: No data, Le present
      this.data = null
      this.le = bytes[4].toInt() and 0xFF
    } else {
      val lc = bytes[4].toInt() and 0xFF
      
      if (bytes.size == 5 + lc) {
        // Case 3: Data present, no Le
        this.data = bytes.copyOfRange(5, 5 + lc)
        this.le = null
      } else if (bytes.size == 5 + lc + 1) {
        // Case 4: Data present, Le present
        this.data = bytes.copyOfRange(5, 5 + lc)
        this.le = bytes[5 + lc].toInt() and 0xFF
      } else {
        throw ValidationError(
          "Invalid APDU length",
          "bytes",
          bytes
        )
      }
    }
    
    validate()
  }
  
  /**
   * Validate the APDU
   * @throws ValidationError if the APDU is invalid
   */
  private fun validate() {
    // Validate data length
    if (data != null && data.size > 255) {
      throw ValidationError(
        "Data length cannot exceed 255 bytes",
        "data",
        data
      )
    }
    
    // Validate Le
    if (le != null && (le < 0 || le > 256)) {
      throw ValidationError(
        "Le must be between 0 and 256",
        "le",
        le
      )
    }
  }
  
  /**
   * Convert the APDU to a byte array
   * @return APDU as byte array
   */
  fun toByteArray(): ByteArray {
    val dataLength = data?.size ?: 0
    val hasLe = le != null
    
    // Calculate the total length
    val length = 4 + (if (dataLength > 0) 1 + dataLength else 0) + (if (hasLe) 1 else 0)
    val bytes = ByteArray(length)
    
    // Header
    bytes[0] = cla
    bytes[1] = ins
    bytes[2] = p1
    bytes[3] = p2
    
    var offset = 4
    
    // Data
    if (dataLength > 0) {
      bytes[offset++] = dataLength.toByte()
      data?.copyInto(bytes, offset)
      offset += dataLength
    }
    
    // Le
    if (hasLe) {
      bytes[offset] = (le!! % 256).toByte()
    }
    
    return bytes
  }
  
  override fun toString(): String {
    val sb = StringBuilder()
    sb.append("CommandAPDU: ")
    sb.append(String.format("CLA=%02X, ", cla))
    sb.append(String.format("INS=%02X, ", ins))
    sb.append(String.format("P1=%02X, ", p1))
    sb.append(String.format("P2=%02X", p2))
    
    if (data != null) {
      sb.append(", Data=")
      data.forEach { sb.append(String.format("%02X", it)) }
    }
    
    if (le != null) {
      sb.append(String.format(", Le=%d", le))
    }
    
    return sb.toString()
  }
}